### DISCUSSION TITLE
Registration of users that already have a record in DB (while onboarding)

### DISCUSSION THREAD

**Discussion by saeedesmaili** [fastapi-users/fastapi-users#1416] [Ideas]
In my app, I want to allow visitors to try a few functionalities of the app before creating an account. At some point in the process they will be asked to create an account if they want to store their progress.

To achieve this, I need to create a new User record when someone starts this process, so their activity can be retained on their profile if they decide to create an account. Is this possible to do with fastapi-users? Any ideas on how to implement this? Here is what I'm thinking of:

1. Create a temporary user record: When a visitor starts using the app, create a new User record with a flag indicating it's a temporary account. I can do this without requiring the usual registration process.
2. Associate actions with the temporary user: Use this temporary user record to associate the visitor's actions and progress.
3. Convert to a permanent account: When the user decides to create a full account, update the existing temporary user record with the new information instead of creating a new one.

But I'm not sure what would be the best way to implement the step 3, specially with oauth. Any help on this is appreciated.

**Comment by frankie567** [comment]
That's a very specific feature, and will greatly depends on what you want to achieve.

The solution you describe seems to be the most versatile, but probably the most complex. Regarding step 3, in terms of implementation, it'll look like the `/register` route mostly. The difference here is that you'll have to find a way to associate this new account with the temporary user. At first thought, I would say a session cookie could be great for this purpose: the session cookie stores the ID of your temporary user. Then, when you call your "upgrade" API, you'll read that cookie and know that you'll have to create a new user for this temporary user ID.

---

Another solution I can see is to store everything in session data: you don't create any object in your database until the account is created. Then, when the user upgrades, you create both the User record and all the other objects directly linked to it. The advantage with this approach is that you don't end up with "ghost" objects in your DB associated with temporary users that never upgraded their account.

**Reply by saeedesmaili** [reply]
Thanks for taking the time to reply!

The proposed solution of storing it in session data is a good one actually. It never came to my mind. I'll see what would work.
